esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -2252,38 +2252,46 @@ void CFileSprayEx::getDropZoneInfoByIP(double clientVersion, const char* ip, con
         return;
     }
 
-    IArrayOf<IConstTpDropZone> dropZones;
-    queryDropZoneInfo(clientVersion, nullptr, false, dropZones);
-    if (dropZones.empty())
-        throw MakeStringException(ECLWATCH_DROP_ZONE_NOT_FOUND, "No dropzone found.");
-
     bool dzFound = false;
     IpAddress ipAddrToMatch(ip);
-    ForEachItemIn(currentDZIndex, dropZones)
+
+    Owned<IConstDropZoneInfoIterator> dropZoneItr = constEnv->getDropZoneIterator();
+    ForEach(*dropZoneItr)
     {
-        IConstTpDropZone& dz = dropZones.item(currentDZIndex);
-        IArrayOf<IConstTpMachine>& servers = dz.getTpMachines();
-        ForEachItemIn(currentServerIndex, servers)
+        IConstDropZoneInfo& dropZoneInfo = dropZoneItr->query();
+
+        SCMStringBuffer dropZoneDirectory, dropZoneUMask;
+        dropZoneInfo.getDirectory(dropZoneDirectory);
+        dropZoneInfo.getUMask(dropZoneUMask);
+        if (!dropZoneDirectory.length())
+            continue;
+
+        Owned<IConstDropZoneServerInfoIterator> machineItr = dropZoneInfo.getServers();
```

Comment:
You have basically implemented it here (you're iterating DZ and then Server and matching IP) and elsewhere.. but it's mixed up with other code.
It would be better to implement it cleanly now and rest of code would be/would have been simpler.